modules/nlu/src/backend/engine2.ts

```diff
@@ -0,0 +1,1240 @@
+  let totalWeight = 0
+  let sentenceEmbedding = new Array(utterance.tokens[0].vectors.length).fill(0)
+
+  for (const token of utterance.tokens) {
+    const norm = computeNorm(token.vectors as number[])
+    if (norm <= 0) {
+      continue
+    }
+    totalWeight += token.tfidf
+    const weightedVec = scalarDivide(token.vectors as number[], norm / token.tfidf)
+    sentenceEmbedding = vectorAdd(sentenceEmbedding, weightedVec)
+  }
+
+  return scalarDivide(sentenceEmbedding, totalWeight)
+}
+
+export const trainIntentClassifer = async (input: TrainOutput, tools: Tools): Promise<_.Dictionary<string>> => {
+  const svmPerCtx: _.Dictionary<string> = {}
+  for (const ctx of input.contexts) {
+    const points = _.chain(input.intents)
+      .filter(i => i.contexts.includes(ctx))
+      .flatMap(i =>
+        i.utterances.map(utt => ({
+          label: i.name,
+          coordinates: computeSentenceEmbedding(utt)
+        }))
+      )
+      .value()
+
+    const svm = new tools.mlToolkit.SVM.Trainer()
+    svmPerCtx[ctx] = await svm.train(points, SVM_OPTIONS, p => debugIntentsTrain('svm progress ==> %d', p))
+  }
+
+  return svmPerCtx
+}
+
+export const trainContextClassifier = async (input: TrainOutput, tools: Tools): Promise<string> => {
+  const points = _.flatMapDeep(input.contexts, ctx => {
+    return input.intents
+      .filter(intent => intent.contexts.includes(ctx) && intent.name !== NONE_INTENT)
+      .map(intent =>
+        intent.utterances.map(utt => ({
+          label: ctx,
+          coordinates: computeSentenceEmbedding(utt)
+        }))
+      )
+  })
+
+  const svm = new tools.mlToolkit.SVM.Trainer()
+  return svm.train(points, SVM_OPTIONS, p => debugIntentsTrain('SVM => progress for CTX %d', p))
+}
+
+export const ProcessIntents = async (
+  intents: Intent<string>[],
+  languageCode: string,
+  list_entities: ListEntityModel[],
+  tools: Tools
+): Promise<Intent<Utterance>[]> => {
+  return Promise.map(intents, async intent => {
+    // TODO filter out non trainable intents (see engine 1 filtering conditions)
+    const cleaned = intent.utterances.map(replaceConsecutiveSpaces)
+    const utterances = await Utterances(cleaned, languageCode, tools)
+
+    const allowedEntities = _.chain(intent.slot_definitions)
+      .flatMap(s => s.entities)
+      .uniq()
+      .value() as string[]
+
+    const entityModels = _.intersectionWith(list_entities, allowedEntities, (entity, name) => {
+      return entity.entityName === name
+    })
+
+    const vocab = buildIntentVocab(utterances, entityModels)
+
+    return { ...intent, utterances: utterances, vocab, slot_entities: allowedEntities }
+  })
+}
+
+export const ExtractEntities = async (input: TrainOutput, tools: Tools): Promise<TrainOutput> => {
+  for (const intent of input.intents) {
+    intent.utterances.forEach(async utterance => await extractUtteranceEntities(utterance, input, tools))
+  }
+
+  return input
+}
```

Comment:
It looks like this is the only method mutating "TrainOutput" (while others make a copy). Shouldn't it be immutable ?

Reply:
yup